src/lib/dma-trace.c

```diff
@@ -375,13 +375,13 @@ static int dtrace_calc_buf_overflow(struct dma_trace_buf *buffer,
 	uint32_t overflow_margin;
 	uint32_t overflow = 0;
 
-	margin = buffer->end_addr - buffer->w_ptr;
+	margin = dtrace_calc_buf_margin(buffer);
 
 	/* overflow calculating */
 	if (buffer->w_ptr < buffer->r_ptr)
-		overflow_margin = buffer->r_ptr - buffer->w_ptr;
+		overflow_margin = buffer->r_ptr - buffer->w_ptr - 1;
```

Comment:
Sorry, I'm not following the commit message. What is the bug this is fixing and why does -1 fix it ? Thanks

Reply:
Occasionally, we got kernel message about dma_trace_buffer_overflow (on yorp). So there still was a case in which we add to many traces to buffer. We calculate overflow in trace_work_function:
```
/* make sure we don't write more than buffer */
	if (avail > DMA_TRACE_LOCAL_SIZE) {
		overflow = avail - DMA_TRACE_LOCAL_SIZE;
		avail = DMA_TRACE_LOCAL_SIZE;
	} else {
		overflow = 0;
	}
```
In my opinion it is caused by invalid overflow_margin calculation. 
In our dma trace buffer we've got particular case in which w_ptr is equal to r_ptr, which means that buffer is empty, NOT that buffer is full. So, the actual buffer length is DMA_TRACE_LOCAL_SIZE-1. During adding traces we should't lead to situation when w_ptr is equal to r_ptr, because it would suggest that buffer is empty.

Reply:
Maybe I will try to give a little example. I will assume that we are working on code before my PR.

Also for simplicity i will assume that we have following situation:
buffer_len = 10 (start_addr = 0; end_addr = 10)
w_ptr = 3
r_ptr = 7
avail = 6 (buffer_len - r_ptr + w_ptr)
overflow_margin = 4 (r_ptr - w_ptr)

Next, we try to invoke dtrace_add_event function with length equal = 4. We update our parameters:
w_ptr = w_ptr + length = 7
r_ptr = 7
avail = avail + 4 = 10

Until then, everything seems to be correct, but... in the next invocation of dtrace_add_event with e.g. length=1 we try to calculate overflow_margin and we get:
overflow_margin = margin + buffer->r_ptr - buffer->addr;
	margin = buffer->end_addr - buffer->w_ptr = 10-7 = 3
overflow_margin = 3 + 7 - 0 = 10

So overflow_margin suggest that buffer is empty, so we add new trace and upgrade avail parameter:
avail = avail + length = 10 + 1 = 11

Next in trace_work we calculate overflow: 
```
if (avail > DMA_TRACE_LOCAL_SIZE) {
		overflow = avail - DMA_TRACE_LOCAL_SIZE;
		avail = DMA_TRACE_LOCAL_SIZE;
	} else {
		overflow = 0;
	}
```
In our case avail is bigger than buffer_len, so we have overflow.